Stop using global initial styles for stylo; the initial styles need to be per-document

bzbarsky opened the pull request:
<!-- Thank you for contributing to Servo! Please replace each `[ ]` by `[X]` when the step is complete, and replace `__` with appropriate data: -->
- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix https://bugzilla.mozilla.org/show_bug.cgi?id=1298588

<!-- Either: -->
- [ ] There are tests for these changes OR
- [X] These changes do not require tests on the servo side because behavior is unchanged.  Gecko-side tests probably exist.

<!-- Pull requests that do not address these steps are welcome, but they will require additional verification as part of the review process. -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/14848)
<!-- Reviewable:end -->

bzbarsky committed:
Bug 1298588 part 2, servo piece.  Pass through an nsPresContext to the PerDocumentStyleData constructor.  r=bholley

bzbarsky committed:
Bug 1298588 part 3, servo piece.  Add a default ComputedValues member to PerDocumentStyleData.  r=bholley

bzbarsky committed:
Bug 1298588 part 4, servo piece.  Recreate the default computed values for a document as needed.  r=bholley

bzbarsky committed:
Bug 1298588 part 5.  Store a ComputedValues on SharedStyleContext.  r=bholley

bzbarsky committed:
Bug 1298588 part 6, servo piece.  Stop using initial_values when doing inheritance in Gecko glue code.  r=bholley

bzbarsky committed:
Bug 1298588 part 7, servo piece.  Stop using initial_values in general in Gecko glue code.  r=bholley

bzbarsky committed:
Bug 1298588 part 8.  Pass a SharedStyleContext, not a Stylist, to Legalizer methods.  r=bholley

bzbarsky committed:
Bug 1298588 part 9, servo piece.  Pass through useful default styles to cascade().  r=bholley

bzbarsky committed:
Bug 1298588 part 10, servo piece.  Pass through useful default styles to apply_declarations().  r=bholley

bzbarsky committed:
Bug 1298588 part 11.  Pass through useful default styles to CascadePropertyFn.  r=bholley

bzbarsky committed:
Bug 1298588 part 12.  Compile some bits that call ComputedValues::initial_values only for servo, not stylo.  r=bholley

Stylist::set_device seems to only be used in servo code, and is the only consumer of ViewportConstraints::maybe_new.

bzbarsky committed:
Bug 1298588 part 13.  Make sure Device has a ComputedValues for stylo. r=bholley

bzbarsky committed:
Bug 1298588 part 14.  Remove ComputedValues::initial_values for stylo.  r=bholley

bzbarsky committed:
Bug 1298588 part 15, servo piece.  Rip out the initial() methods on style structs in stylo.  r=bholley

highfive commented:
Heads up! This PR modifies the following files:
 * @bholley: components/style/gecko_bindings/bindings.rs, components/style/gecko/data.rs, components/style/gecko_bindings/structs_debug.rs, components/style/build_gecko.rs, components/style/gecko_bindings/structs_release.rs, components/style/media_queries.rs, components/style/properties/gecko.mako.rs, components/style/values/computed/mod.rs, components/style/properties/properties.mako.rs, components/style/properties/helpers.mako.rs, components/style/viewport.rs, components/style/stylist.rs, components/style/animation.rs, components/style/context.rs, components/style/matching.rs
 * @kichjang: components/script_layout_interface/wrapper_traits.rs
 * @fitzgen: components/script_layout_interface/wrapper_traits.rs
 * @emilio: components/style/gecko_bindings/bindings.rs, components/style/gecko/data.rs, components/style/gecko_bindings/structs_debug.rs, ports/geckolib/glue.rs, components/style/build_gecko.rs, components/style/gecko_bindings/structs_release.rs, components/style/media_queries.rs, components/style/properties/gecko.mako.rs, components/style/values/computed/mod.rs, components/style/properties/properties.mako.rs, components/layout/construct.rs, components/style/properties/helpers.mako.rs, components/style/viewport.rs, components/style/stylist.rs, components/style/animation.rs, components/style/context.rs, components/style/matching.rs

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify style and layout code, but no tests are modified. Please consider adding a test!

bholley commented:
@bors-servo delegate+

Manishearth commented:
@bors-servo r=bholley

bors-servo commented:
:pushpin: Commit 3e899da has been approved by `bholley`

<!-- @bors-servo r=bholley 3e899dac412d74a764f4970ab40e4665e7ed28c2 -->

bors-servo commented:
:hourglass: Testing commit 3e899da with merge bb5110e...

bors-servo commented:
:broken_heart: Test failed - [linux-dev](http://build.servo.org/builders/linux-dev/builds/5018)

bzbarsky commented:
I'm seeing the following test failures, repeated a few times:

1. test_3 build failure due to using Context::initial, which no longer exists.  Fixed.
2. build-geckolib failure.  This is a merge issue: we started requiring comments on things, but my comment (which I did have!) started with only two '/', which is apparently not good enough.  Fixed.

bzbarsky commented:
@bors-servo r=bholley

bors-servo commented:
:pushpin: Commit dd80b5c has been approved by `bholley`

<!-- @bors-servo r=bholley dd80b5c0e7b0ad62b7897ac17782d573a2fa16ea -->

bors-servo commented:
:hourglass: Testing commit dd80b5c with merge 143dfc8...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/4725), [arm32](http://build.servo.org/builders/arm32/builds/4894), [arm64](http://build.servo.org/builders/arm64/builds/4928), [linux-dev](http://build.servo.org/builders/linux-dev/builds/5019), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/1696), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/1702), [mac-dev-unit](http://build.servo.org/builders/mac-dev-unit/builds/4996), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/5035), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/1702), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/1738), [windows-gnu-dev](http://build.servo.org/builders/windows-gnu-dev/builds/158), [windows-msvc-dev](http://build.servo.org/builders/windows-msvc-dev/builds/159)
